### PR TITLE
[API] As a user, I can login

### LINKED ISSUE #70: [API] As a user, I can login
### Why?
User can login with an API by using oauth client with their credential for requesting an access token

### Acceptance Criteria
- [x] Provide and endpoint [POST]`v1/oauth/token`
- [x] Authenticate the user's credential
- [x] Return an access token

### PR DESCRIPTION
**Pull request by Lahphim**
Resolve https://github.com/Lahphim/go-crawler-challenge/issues/70

## What happened 👀

Provide an endpoint `/api/v1/oauth/token` for creating a new token information binds with the current user credentials ✨

## Insight 📝

- [x] Provide and endpoint [POST]v1/oauth/token
- [x] Authenticate the user's credentials
- [x] Return an access token

Extra:

- Set up [oauth2](https://github.com/go-oauth2/oauth2) and [go-oauth2-pg](https://github.com/vgarvardt/go-oauth2-pg) for managing about `client` and `token`
- Feel free to check the API documentation here: https://documenter.getpostman.com/view/105704/TzCMc7RN
- Following this PR to see how to generate a new client credentials: https://github.com/Lahphim/go-crawler-challenge/pull/69
- Override `oauth2` response following `JSON-LD` method

## Proof Of Work 📹

[POST] /api/v1/oauth/token
Request a new token information
| Valid | Invalid |
|------|-------|
| Screen Shot 2564-03-31 at 18 06 58 | Screen Shot 2564-03-31 at 18 08 09 |

### CODE REVIEW

File: conf/initializers/oauth2.go [hunk 1]
```diff
@@ -61,10 +61,7 @@ func SetUpOauth2() {
 func internalErrorHandler(err error) (response *errors.Response) {
```
Comment (olivierobert) [inline conf/initializers/oauth2.go]: Shouldn't the return type be removed as well since now nothing is returned by the method? 🤔

Comment (Lahphim) [reply conf/initializers/oauth2.go]: @olivierobert Removed all unused handler ✌️

**Review by junan — APPROVED** [review]

File: controllers/api/base.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+func (c *BaseController) RenderUnauthorizedError(err error) {
+	statusCode := http.StatusUnauthorized
+
+	c.RenderError(http.StatusText(statusCode), err.Error(), statusCode, "unauthorized_error")
+}
+
+func (c *BaseController) RenderError(title string, detail string, status int, code string) {
+	c.Ctx.Output.Header("Content-Type", ContentType)
```
Comment (hoangmirs) [inline controllers/api/base.go:48]: Should we set the `Content-Type` header in `Prepare()` method? Then we can set it for all requests

File: controllers/api/v1/token_test.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+				var responseToken v1serializers.TokenInformationResponse
+				err := jsonapi.UnmarshalPayload(response.Body, &responseToken)
+				if err != nil {
+					Fail(fmt.Sprintf("Unmarshal payload `TokenInfo` failed: %v", err.Error()))
+				}
+
+				Expect(len(responseToken.AccessToken)).To(BeNumerically(">", 0))
+				Expect(len(responseToken.RefreshToken)).To(BeNumerically(">", 0))
+				Expect(len(responseToken.AccessToken)).To(BeNumerically(">", 0))
```
Comment (hoangmirs) [inline controllers/api/v1/token_test.go]: Is this one duplicated?

File: controllers/api/base.go (hunk 2 above)
Comment (Lahphim) [reply controllers/api/base.go:48]: @hoangmirs I'll update in this PR https://github.com/Lahphim/go-crawler-challenge/pull/73/commits/b05f6b9cc322295a302bf70a640742f78eda5beb

**Review by hoangmirs — APPROVED** [review]

**Pull request merged**